Fixes #545 Create Photo Gallery Paragraph Type

tadean opened the pull request:
This pull request adds a dual purpose Photo Gallery paragraph that covers the functionality of the old photo gallery and carousel types from QS1.

## Description
This PR contains the following:

- a paragraph module (`az_paragraphs_photo_gallery`) that contains `az_photo_gallery`
- the addition of a `caption` field to the `az_image` media bundle
- a paragraph behavior plugin that allows choosing between the grid display and the slider display as an option.
- a template that handles rendering both cases; the grid uses a modal display with hint images, and the slider display uses a carousel.

## Related Issue
#545 

## How Has This Been Tested?
- Create a Flexible Page
- Add a Photo Gallery paragraph
- Add several media items to the paragraph
- Choose a display type, eg. Slider or Grid.
- Save content
- Verify functionality of the Photo Gallery based on type:
- Grid should open a Modal when previews are clicked on.
- Grid modal's close button should operate.
- Slider's carousel controls should operate.
- Verify that the gallery continues to function in preview mode (on the edit page)
- Verify that multiple galleries can co-exist on the same page.
- Verify that other content types/paragraphs are not impacted by the addition of the caption field, eg. cards, news, etc.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tadean commented:
Two galleries, 1 of the slider type (top), 1 of the grid type(bottom):
![image](https://user-images.githubusercontent.com/51835841/114049123-07231280-9840-11eb-94f5-27ac9581f708.png)

Grid type when open:
![image](https://user-images.githubusercontent.com/51835841/114049174-173af200-9840-11eb-839a-389df2decba4.png)

tadean commented:
The **Security workflow / Dependency Check (push)** step appears to have encountered a web service error when reaching the registry, unrelated to this PR.

akslay commented:
Just as a conversation item for the group: did we want to have a more uniform height/max-height on the slideshow version (when not in modal mode) as an improvement to UX? When you have a combination of landscape and portrait orientation images there's a lot of jumping on the page since the height varies. Which also causes the arrows to move around. This is how it currently works in QS 1.0, but it might be a good improvement for 2.0.

Maybe as a follow-up improvement (if other people like the idea of a uniform height) for GA rather than having it included in this ticket.

tadean commented:
> * What do people think about having a more uniform height/max-height on the slideshow version (when not in modal mode) as an improvement to UX? When you have a combination of landscape and portrait orientation images there's a lot of jumping on the page since the height varies. Which also causes the arrows to move around. This is how it currently works in QS 1.0, but it might be a good improvement for 2.0.

I'm curious about this too. In this PR I implemented it how it behaves in QS 1 with regard to the height changing, but I find the QS1 jumping behavior a little off putting. From a code standpoint it'd certainly be possible to enforce a height on the slider images if we choose to do so.

trackleft commented:
@akslay do you think we should update bootstrap to do that, or crop images in QS to do that?

smayden commented:
Super excited for this PR! Initial testing is looking great. Thanks @tadean 

A few questions to consider for now and/or future improvements:

-In QS1 for smaller screens we to do 2 images per row, do we want to do that here for the grid display? (I'm seeing one image per row on my phone)
-Is there a way to get back to edit a caption for media already added through the editing interface (perhaps future improvement and there may already be a ticket for this)
-Images in QS1 have a separate photo credit field that is probably useful for galleries. Possibly add for future (I think there is a ticket for creating field separately)
-How long is the caption field. I tried to recreate one of the galleries we have on the UANews site and some captions were too long to fit: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-650.probo.build/gallery-gov-ducey-visits-uarizona-vaccination-site-vaccine-eligibility-expands
-QS1 has an indicator for photo number as you through the gallery modal (e.g. 1 of 8). Is it helpful/easy to add that in future?
-The arrows on the slide were hard to see for me on some images (blending into background, small size). Possible accessibility issue?

![image](https://user-images.githubusercontent.com/10873398/114091621-c3470200-986d-11eb-8352-67fd52a79a95.png)

tadean commented:
> Super excited for this PR! Initial testing is looking great. Thanks @tadean

Thanks @smayden !

> -Is there a way to get back to edit a caption for media already added through the editing interface (perhaps future improvement and there may already be a ticket for this)

They can currently be edited again through the media interface, at `/admin/content/media` It'd be nice if the media reference widget allowed doing this in the future.

> -How long is the caption field. I tried to recreate one of the galleries we have on the UANews site and some captions were too long to fit: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-650.probo.build/gallery-gov-ducey-visits-uarizona-vaccination-site-vaccine-eligibility-expands

 The length limit of the Caption field is currently 255 characters - how long should we allow?

smayden commented:
@tadean wrote:

> The length limit of the Caption field is currently 255 characters - how long should we allow?

I think it QS1 the caption field is a text area that does not have a character limit. Curious to hear other opinions on this as the news site is probably more of an outlier when it comes to crazy long captions.

tadean commented:
> @tadean wrote:
> 
> > The length limit of the Caption field is currently 255 characters - how long should we allow?
> 
> I think it QS1 the caption field is a text area that does not have a character limit. Curious to hear other opinions on this as the news site is probably more of an outlier when it comes to crazy long captions.

In QS1 the two original paragraphs this new "combined" paragraph was based on had different lengths allowed. The photo gallery was a text box with unlimited entry (through the media interface), and the carousel captions allowed only 255.

akslay commented:
> @akslay do you think we should update bootstrap to do that, or crop images in QS to do that?

@trackleft I'm not sure where the best place for that would be. 🤔 I think I personally might lean more towards having it on the AZQS side rather than the Bootstrap side. Because I see AZQS as more "tailored" product and see Bootstrap more of a tool to help you build something tailored. But I wouldn't have any qualms with having it on one or the other.

danahertzberg commented on modules/custom/az_paragraphs/az_paragraphs_photo_gallery/templates/paragraph--az-photo-gallery.html.twig:
```diff
@@ -0,0 +1,132 @@
+              '#alt':        item.entity.field_media_az_image.0.entity.uri.alt,
+              '#attributes': { class: [ 'photo-gallery-grid-img' ] },
+            } %}
+            <div class="col-xs-6 col-sm-6 col-md-4 col-lg-3 px-min py-min" data-toggle="modal" data-target="#{{ modal }}">
```
```suggestion
            <div class="col-6 col-md-4 col-lg-3 px-min py-min" data-toggle="modal" data-target="#{{ modal }}">
```

danahertzberg commented:
Follow up tasks: https://github.com/az-digital/arizona-bootstrap/issues/265

tadean commented:
Updated with the following changes:

- Made the minimum column width of the gallery previews `col-6` and removed `col-xs-6` and `col-sm-6`
- Changed the caption field to be a text long rather than text. Format is still set to formatted text, plain text.

smayden reviewed (approved):
Tested again as a content editor. Can add longer captions now and verified that I see grid with two photos per row on mobile.

tadean commented:
> Tested again as a content editor. Can add longer captions now and verified that I see grid with two photos per row on mobile.

Thanks @smayden !
